File: 259_3sum_smaller.java
class Solution {
    // Time Complexity: O(n * twoSumSmalller) = O(n ^ 2)
    // Space Complexity: O(1)
    public int threeSumSmaller(int[] nums, int target) {
        int sum = 0;
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            sum += twoSumSmaller(nums, i + 1, target - nums[i]);
        }
        return sum;
    }

    // Time Complexity: O(n) (because two pointer with n steps)
    // Space Complexity: O(1)
    private int twoSumSmaller(int[] nums, int startIdx, int target) {
        int sum = 0;
        int left = startIdx, right = nums.length - 1;
        while (left < right) {
            if (nums[left] + nums[right] < target) {
                sum += right - left;
                left++;
            } else {
                right--;
            }
        }
        return sum;
    }
}

File: python/259_3sum_smaller.py
class Solution:
    # Time Complexity: O(n ^ 2)
    # Space Complexity: O(1)
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        nums.sort()
        count = 0
        for i in range(len(nums) - 2):
            # no need to consider duplication since we want all results
            low, high = i + 1, len(nums) - 1
            # if current sum < target, increase low
            # otherwise, decrease high
            while low < high:
                curr_sum = nums[i] + nums[low] + nums[high]
                if curr_sum < target:
                    count += (high - low)
                    low += 1
                else:
                    high -= 1
        return count
